KAFKA-6774: Improve the default group id behavior in KafkaConsumer (KIP-289)

vahidhashemian opened the pull request:
Improve the default group id behavior by
* changing the default consumer group to null, where no offset commit or fetch, or group management operations are allowed
* deprecate the use of empty (`""`) consumer group on the client


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vahidhashemian committed:
KAFKA-6774: Improve the default group id behavior in KafkaConsumer (KIP-289)

Improve the default group id behavior by
* changing the default consumer group to null, where no offset commit or fetch, or group management operations are allowed
* deprecate the use of empty (`""`) consumer group on the client

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2216,6 +2231,11 @@ private void throwIfNoAssignorsConfigured() {
 
+    private void maybeThrowInvalidGroupIdException() {
+        if (groupId == null)
+            throw new InvalidGroupIdException("The null group id does not support group management operations.");
```
Perhaps group management or the the offset commit api?

hachikuji reviewed (commented):
@vahidhashemian Thanks for the PR. I think we also might need to hook into the logic for setting the initial position of the consumer. By default, we attempt to fetch committed offsets, but if the group id is null, we should just skip this step.

vahidhashemian committed:
Avoid offset fetch when group id is null

vahidhashemian commented:
@hachikuji Thanks for reviewing the PR. Submitted a patch to address your feedback.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -662,10 +664,15 @@ private KafkaConsumer(ConsumerConfig config,
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            if (groupId == null)
+                // overwrite in case of default group id
+                enableAutoCommit = false;
```
I think it's fine to change the default as long as the user hasn't explicitly provided the configuration. However, if they have explicitly set `enable.auto.commit=true`, then it might be better to raise an exception.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -1271,6 +1281,57 @@ public void closeShouldBeIdempotent() {
+    @Test
+    public void testOperationsBySubscribingConsumerWithDefaultGroupId() {
+        try {
+            newConsumer((String) null).subscribe(Collections.singleton(topic));
```
We should probably have a `fail` below if this returns without errors?

vahidhashemian replied:
Thanks! Totally missed them.

hachikuji reviewed (commented):
Left a couple more comments. Do we have an integration test that covers basic fetching using a null group id? We should probably make sure we have at least one case which uses null and one which uses the empty string.

vahidhashemian committed:
Added a couple of integration tests

Also modified the logic in deciding whether auto-commit should be enable.

vahidhashemian reviewed (commented):
@hachikuji I tried to address your recent comments in the new commit. Thanks for the additional feedback.

vahidhashemian committed:
Minor unit test updates based on the recent feedback

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    consumer1Config.put(ConsumerConfig.CLIENT_ID_CONFIG, "consumer1")
+    val consumer1 = createConsumer(
+      configOverrides = consumer1Config,
+      configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
```
Is it possible to just pass "groupId=null" as an override?

vahidhashemian replied:
Doesn't seem so, I get a NPE:
```
java.lang.NullPointerException
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1021)
	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1016)
	at java.base/java.util.Properties.put(Properties.java:1309)
	at kafka.api.PlaintextConsumerTest.testConsumingWithNullGroupId(PlaintextConsumerTest.scala:1838)
        ...
```

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+
+    // consumer 1 uses the default group id and consumes from earliest offset
+    val consumer1Config = new Properties(consumerConfig)
+    consumer1Config.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
```
I suppose we could leave this out?

vahidhashemian replied:
That's correct.

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+      configOverrides = consumer2Config,
+      configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
+
+    // consumer 3 uses the default group id and consumes from earliest offset
```
Seems more accurate to say that consumer 3 begins from an explicit seek?

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    consumer3.close()
+
+    assertTrue("Expected consumers to consumer a single message each from the last committed offset",
+      records1.count() == 1 && records1.records(tp).asScala.head.offset == 0 &&
```
nit: maybe better to turn this into explicit equals assertions so that we know which check failed.

vahidhashemian replied:
Sounds good.

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    val records2 = consumer2.poll(Duration.ofMillis(5000))
+    consumer2.commitSync()
+
+    val records3 = consumer3.poll(Duration.ofMillis(5000))
```
The third iteration here feels redundant. Doesn't it just read from the committed offset like `consumer2`?

vahidhashemian replied:
Yes, I put it to verify the `auto.offset.reset` doesn't affect the behavior. I'll remove it in the new commit.

hachikuji reviewed (commented):
Thanks for the tests. Left a few minor comments.

vahidhashemian reviewed (commented):
@hachikuji I updated the unit tests based on your comments. Thanks!

hachikuji reviewed (commented):
@vahidhashemian Thanks for the patch. Looks good overall. I had just one thought. When the groupId is null, there should be no coordinator interaction at all. We could make this explicit by not bothering to construct the `ConsumerCoordinator` object. That would take it out of the poll path and reduce the risk of using it accidentally. What do you think?

vahidhashemian committed:
Disable coordinator construction when group id is null

vahidhashemian commented:
@hachikuji thanks for the feedback. Your suggestion makes sense. Even though it's not explicitly mentioned in the KIP, I think it wouldn't result in a different behavior that what is covered by the KIP. I submitted a patch that should cover it (it did not break any test locally), but please let me know if I missed something. Thanks!

vahidhashemian committed:
Additional minor fixes

vahidhashemian committed:
Remove group id validation from `unsubscribe()` method

The validation caused Streams tests in `KTableSourceTopicRestartIntegrationTest` to fail.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -491,6 +491,9 @@ public boolean rejoinNeededOrPending() {
      */
     public boolean refreshCommittedOffsetsIfNeeded(Timer timer) {
+        // no committed offset should be fetched for the default group id
+        if (groupId == null) return true;
```
Perhaps now we don't need this? We could add an assertion in the constructor that the groupId is not null.

vahidhashemian replied:
You're right. I added an assertion in `AbstractCoordinator` constructor. It seemed to be a better fit there.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -203,7 +206,7 @@ public void testInvalidSocketReceiveBufferSize() {
     @Test
     public void testSubscription() {
-        KafkaConsumer<byte[], byte[]> consumer = newConsumer();
+        KafkaConsumer<byte[], byte[]> consumer = newConsumer("");
```
nit: not sure if there's a strong reason for any of these to use the empty group id. Could we use `groupId` instead?

vahidhashemian replied:
No particular reason other than I wanted to stick to the actual group id that was used before. I updated these to use the provided `groupId`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2216,6 +2228,11 @@ private void throwIfNoAssignorsConfigured() {
 
+    private void maybeThrowInvalidGroupIdException() {
+        if (groupId == null)
+            throw new InvalidGroupIdException("The null group id does not support group management or offset commit APIs.");
```
I was wondering if we could rephrase this to suggest what the user should do more clearly. For example, "To use the group management or offset commit APIs, you must provide a valid `group.id` in the consumer configuration."

vahidhashemian replied:
Sounds good, thanks!

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1814,4 +1815,110 @@ class PlaintextConsumerTest extends BaseConsumerTest {
   }
 
+  @Test
+  def testConsumingWithNullGroupId(): Unit = {
```
Just to complete the integration picture, could we make a call to `committed` and assert the exception?

vahidhashemian replied:
I did that in the next commit.

hachikuji reviewed (commented):
Thanks for the updates, just a few additional comments.

vahidhashemian reviewed (commented):
@hachikuji thanks for the feedback/suggestions. The new commit should address them.

hachikuji commented:
@vahidhashemian Can you look into the build failures? I looked at one of the streams tests (`KTableSourceTopicRestartIntegrationTest`)  and saw there was some dependence on the empty group id.
```
Exception in thread "ktable-restore-from-source-2639ce74-b612-423d-aaff-96cc59653aa7-StreamThread-1" org.apache.kafka.streams.errors.StreamsException: stream-thread [ktable-restore-from-source-2639ce74-b612-423d-aaff-96cc59653aa7-StreamThread-1] Failed to rebalance.
	at org.apache.kafka.streams.processor.internals.StreamThread.pollRequests(StreamThread.java:922)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:822)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:777)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:747)
Caused by: org.apache.kafka.common.errors.InvalidGroupIdException: To use the group management or offset commit APIs, you must provide a valid group.id in the consumer configuration.
```
cc @guozhangwang

vahidhashemian commented:
Thanks @hachikuji for catching this. 
I see this code in `StreamsConfig.getRestoreConsumerConfigs`:
```
        // no need to set group id for a restore consumer
        baseConsumerProps.remove(ConsumerConfig.GROUP_ID_CONFIG);
```
which now means setting the group id to null. I think it should be safe to replace it with
```
        // no need to use a specific group id for a restore consumer
        baseConsumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "");
```
Thoughts?
@guozhangwang

hachikuji commented:
The exception is being raised from a call to `unsubscribe`. We currently allow this API to be used in order to remove all assigned partitions, so I think we should probably remove the group id validation from that API.

vahidhashemian commented:
Sounds good, and seems to have no side effect with respect to existing unit tests. My initial thought was to use the empty string to keep the tests run with the same config as before. I'll submit a patch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -654,19 +656,23 @@ public KafkaConsumer(Properties properties,
 
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            if (groupId == null) { // overwrite in case of default group id where the config is not explicitly provided
+                if (!config.originals().containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG))
```
Shouldn't we handle the `else` case? I was thinking that if the user enables auto commit explicitly, but provides no group id, then we should raise a configuration exception.

vahidhashemian replied:
What I had here was to let the consumer run and throw an exception when a commit is triggered (or even earlier when e.g. `subscribe` is called). But I think your suggestion makes more sense because it points directly to the root cause, i.e. the misconfiguration. Updated in the new patch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -136,6 +136,7 @@ public AbstractCoordinator(LogContext logContext,
                                Time time,
                                long retryBackoffMs,
                                boolean leaveGroupOnClose) {
+        assert groupId != null : "Expected a non-null group id for coordinator construction";
```
nit: can we use `Objects.requireNonNull` instead?

vahidhashemian replied:
Made this change too.

hachikuji reviewed (commented):
Thanks for the updates. Looks like the build is passing. Just two more comments and I think we can merge.

vahidhashemian committed:
Throw a config exception for default group id with auto commit

vahidhashemian reviewed (commented):
@hachikuji Thanks for the feedback. Responded to your comment inline and submitted a new patch.

hachikuji committed:
Fix requireNonNull check

hachikuji reviewed (approved):
LGTM. Pushed a trivial fix for the null check in `AbstractCoordinator`.
